### DISCUSSION TITLE
wire:target exception

### DISCUSSION THREAD

**Discussion by amiranagram** [livewire/livewire#3248] [3. Feature ideas] [answered]
I'd really like the option where instead of listing all the targets that `wire:loading` should be triggered on there was something like `wire:target.except="action"` where it would trigger `wire:loading` DOM manipulations on all actions except the one flagged in this potential directive.

I have this in one of my projects:

```
<div
    wire:loading.class="opacity-40"
    wire:key="invoice-{{ $invoice->timestamp }}"
    wire:target="filter,resetFilters,perPage,sortBy,previousPage,nextPage,gotoPage,resetPage,deleteUser"
>...</div>
```

and it gets repeated a bunch of times in a bunch of places.

What do you guys think?

**Comment by amiranagram** [comment]
Hi I managed to do this on my fork of Livewire. Check out this commit:

https://github.com/amiranagram/livewire/commit/db99b5e5ce202056c4a68c31e9600622ada4e083

Currently doesn't work on models, but I can probably get it working and PR this if feature is warranted.

**Comment by joshhanley** [comment]
@amiranagram I haven't had a need for this (yet 😆), but could be useful!

Can you explain your use case a bit more, such as if you were to use except, how would that work? What sort of action would you want ignored for example? And how would that look on a page?

Thanks! 🙂

**Reply by amiranagram** [reply]
Well, in complex Livewire components there would naturally be a bunch of methods (actions) for which we could target loading states. `wire:target.except` would target loading states for all actions except those defined in the attribute.

For example take the datatable component (not very complex, but large with tons of actions) that Caleb has built in his screencast series (https://laravel-livewire.com/screencasts/s7-intro). He has written a loading state for table rows that toggles their opacity.

```html
<tr wire:loading.class="opacity-50">
    ...
</tr>
```

Now imagine I have an export button with a spinner icon that shows only when that button is clicked:

```html
<button wire:click="exportData">
    <loading-spinner-icon wire:loading wire:target="exportData" />
    Export
</button>
```

Now, back to the table rows where we are toggling opacity, If I don't want to toggle it when pressing the export button I'd have to go and look at all the actions that I wanna trigger it for. It would look something like this:

```html
<tr wire:loading.class="opacity-50" wire:target="filter,resetFilters,perPage,sortBy,previousPage,nextPage,gotoPage,resetPage">
    ...
</tr>
```

With `wire:target.except` (or name it however you want) you could write a lot less HTML and also save time looking up the actions you wanna target. Finally it would look like this:

```html
<tr wire:loading.class="opacity-50" wire:target.except="exportData">
    ...
</tr>
```

Where naturally loading state will be applied on all actions except `exportData`.

Hope I demonstrated it clearly. @joshhanley

**Reply by joshhanley** [reply]
@amiranagram yep makes sense to me. I asked Caleb on my PR to have a look at this, and if he's happy with it, I will have a look at how hard it will be to implement.

**Reply by amiranagram** [reply]
Thanks @joshhanley. Feel free to look at my fork, I hope it saves you some time.

https://github.com/amiranagram/livewire/commit/db99b5e5ce202056c4a68c31e9600622ada4e083

**Comment by joshhanley** [comment]
@amiranagram support for targeted model loading states has been added #3530, so you could have an array of data and set the loading state on the base array if any of the data inside it is modified.

Caleb didn't seem to keen on adding the `.except` for now though. So we will need to see if any one else needs this feature before submitting a PR.

**Reply by amiranagram** [reply]
Thanks @joshhanley for your contribution. No worries. 🙂

**Reply by undjike** [reply]
> So we will need to see if any one else needs this feature before submitting a PR.

+1

**Comment by joelwmale** [comment]
I've also got a pretty big use case for this.

Building a job search platform, with tonnes of filters, and the location search relying on an API request to filter down the results from the initial mount method (this could honestly be refactored to be done in alpine for performance, as slow connections take forever to filter, but that's a post launch issue).

So I can't just do a `wire:loading` due to the fact that it makes requests each time the location variable is updated, so typing in this field, will put the page into a loading state, when what is returned, doesn't affect the results until they click "Submit".

But I need to show the loading animation, any time a filter is changed, so now my `wire:target` looks like this:

`wire:target="search, removeFilter, calculateFilterPills, sortBy, setSortBy, typeFilter, annuallyFromFilter, annuallyToFilter, hourlyFromFilter, hourlyToFilter, postedFilter, settingFilter, verifiedStatusFilter, entryLevelFilter, extraFilter"`

Where as it could be:

`wire:target.except="location"`.

I would say this would be useful for anyone who doesn't know alpine too well to do things like filtering initial loaded data on the fly, and relies on doing this in php, and also has lots of filters/other things that change results.

**Comment by MuminovicAnel** [comment]
It would make sense for my case also, it's a very good idea.

I have a language switcher in the header of the home page and it dispatches an event, It's the only one where I don't want to trigger the loading, And also, I don't want to put all of my functions inside the wire:target (because I have a lot of them and also because there are dispatch events where I want to show the loading icon and it doesn't work that way for events).

I hope this request will get some support. :)

**Comment by patriktoth67** [comment, marked as answer]
I did a PR for this if anyone is interested #8065

**Comment by patriktoth67** [comment]
v3.4.9 includes this feature now

**Comment by blazery** [comment]
@amiranagram This issue can be closed as it has been implemented!
